### DISCUSSION TITLE
Key binding to spawn external program

### DISCUSSION THREAD

**Discussion by ECon65** [wezterm/wezterm#5765] [Q&A]
### What Operating System(s) are you running on?

Windows

### WezTerm version

20240203-110809-5046fc22

### Ask your question!

Is it possible to launch external program (like VSCode) using action of certain key binding?

For example, would it be possible to change the action of following sample [Opening whole scrollback in vim](https://wezfurlong.org/wezterm/config/lua/wezterm/on.html?h=wezterm.on#example-opening-whole-scrollback-in-vim)  to use `code -r` to open the scrollback in VSCode as editor instead of Vim?

**Comment by wez** [comment]
Take a look at https://wezfurlong.org/wezterm/config/lua/wezterm/background_child_process.html

**Reply by ECon65** [reply]
Most likely I'm doing something incorrectly. Using the example of editing the scrollback, I've tried to achieve a usecase where nvim is used for editing (in split) when running under linux and vscode when running under windows.


```lua
wezterm.on("trigger-vim-with-visible-text", function(window, pane)
	local dims = pane:get_dimensions()
	local scrollback_text =
		pane:get_text_from_region(0, dims.scrollback_top, 0, dims.scrollback_top + dims.scrollback_rows)

	local name = os.tmpname()
	local f = io.open(name, "w+")
	f:write(scrollback_text)
	f:flush()
	f:close()

	if wezterm.target_triple == "x86_64-pc-windows-msvc" then
		window:perform_action(
			wezterm.action_callback(function(_, pane)
				wezterm.background_child_process({
					"code",
					"-r",
					name,
				})
			end),
			pane
		)
	end

	if wezterm.target_triple == "x86_64-unknown-linux-gnu" then
		window:perform_action(
			act.SplitPane({
				direction = "Down",
				command = {
					args = {
						"nvim",
						"--",
						name,
					},
				},
				size = { Percent = 50 },
			}),
			pane
		)
	end

	wezterm.sleep_ms(1000)
	os.remove(name)
end)
```

**Reply by wez** [reply]
You don't need the perform_action/action_callback for spawning the background process.
Try something like this:

```lua
wezterm.on("trigger-vim-with-visible-text", function(window, pane)
	local dims = pane:get_dimensions()
	local scrollback_text =
		pane:get_text_from_region(0, dims.scrollback_top, 0, dims.scrollback_top + dims.scrollback_rows)

	local name = os.tmpname()
	local f = io.open(name, "w+")
	f:write(scrollback_text)
	f:flush()
	f:close()

	if wezterm.target_triple == "x86_64-pc-windows-msvc" then
				wezterm.background_child_process({
					"code",
					"-r",
					name,
				})
	end

	if wezterm.target_triple == "x86_64-unknown-linux-gnu" then
		window:perform_action(
			act.SplitPane({
				direction = "Down",
				command = {
					args = {
						"nvim",
						"--",
						name,
					},
				},
				size = { Percent = 50 },
			}),
			pane
		)
	end

	wezterm.sleep_ms(1000)
	os.remove(name)
end)
```

**Reply by ECon65** [reply]
Hi Wez, thank you for helping me. 

I've tried your suggestion but vscode is still not opening. I can see that the temporary file is created in the correct place but nothing else is happening. Are there some logs that I can try to check to see why it is failing?

**Reply by ECon65** [reply]
Hi Wez, I was able to find the following entry in debug overlay

If the ":28" refers to a line number then it is the line with "code" argument. I have the vscode bin folder in the env path and also I can run the command "code -r <filename>" from the terminal to open files in vscode. So I would assume that the program should be known.

08:58:12.918 ERROR wezterm_gui::termwindow > while processing trigger-vim-with-visible-text event: program not found
stack traceback:
        [C]: in local 'poll'
        [string "?"]:4: in function 'wezterm.background_child_process'
        C:\Users\<my_user>\.config\wezterm/lua\shortcuts.lua:28: in function <C:\Users\<my_user>\.config\wezterm/lua\shortcuts.lua:7>

**Reply by wez** [reply]
sounds like you need to use the full path to the `code` executable

**Reply by ECon65** [reply]
after providing the full path the error has changed slightly

```lua
	if wezterm.target_triple == "x86_64-pc-windows-msvc" then
		-- Configs for Windows only
		wezterm.background_child_process({
			"C:\\Users\\<my_user>\\AppData\\Local\\Programs\\Microsoft VS Code\\bin\\code",
			"-r",
			name,
		})
	end


20:08:48.253 ERROR wezterm_gui::termwindow > while processing trigger-vim-with-visible-text event: %1 is not a valid Win32 application. (os error 193)
stack traceback:
        [C]: in local 'poll'
        [string "?"]:4: in function 'wezterm.background_child_process'
        C:\Users\<my_user>\.config\wezterm/lua\shortcuts.lua:28: in function <C:\Users\<my_user>\.config\wezterm/lua\shortcuts.lua:7>
```

I was able to fix it with the following change. Not sure whether it is correct solution but it works.
```lua
	if wezterm.target_triple == "x86_64-pc-windows-msvc" then
		-- Configs for Windows only
		wezterm.background_child_process({
			"cmd",
			"/C",
			"code",
			"-r",
			name,
		})
	end
```
